Sources/UIApplication+handleSDLEvents.swift

```diff
@@ -143,3 +150,122 @@ extension SDL_Event {
         return TimeInterval(self.common.timestamp) / 1000
     }
 }
+
+extension UIEvent {
+    static func from(_ event: SDL_Event) -> UIEvent? {
+        switch SDL_EventType(event.type) {
+        case SDL_FINGERDOWN:
+            // XXX: minimal implementation for single touch
+            guard UIEvent.activeEvents.isEmpty else {
+                return nil
+            }
+
+            let newTouch = UITouch(
+                touchId: 0,
+                at: CGPoint(
+                    x: CGFloat(event.tfinger.x),
+                    y: CGFloat(event.tfinger.y)
+                ),
+                timestamp: event.timestampInSeconds
+            )
+
+            return UIEvent(touch: newTouch)
+
+        case SDL_FINGERMOTION:
+            if
+                let firstExistingEvent = UIEvent.activeEvents.first,
+                let matchingTouch = firstExistingEvent.allTouches?.first(where: {$0.touchId == event.tfinger.fingerId})
+            {
+
+                matchingTouch.timestamp = event.timestampInSeconds
+                matchingTouch.phase = .moved
+                matchingTouch.updateAbsoluteLocation(.from(event.tfinger))
+                return firstExistingEvent
+            } else {
+                return nil
+            }
+
+        case SDL_FINGERUP:
+            if
+                let firstExistingEvent = UIEvent.activeEvents.first,
+                let matchingTouch = firstExistingEvent.allTouches?.first(where: {$0.touchId == event.tfinger.fingerId})
+            {
+                matchingTouch.timestamp = event.timestampInSeconds
+                matchingTouch.phase = .ended
+                return firstExistingEvent
+            } else {
+                return nil
+            }
+
+        default:
+            return nil
+        }
+    }
+}
+
+extension UIEvent: CustomStringConvertible {
+    public var description: String {
+        let text = "Event with touches: "
+        guard let allTouches = self.allTouches else {
+            return text + "none"
+        }
+        let descriptions = allTouches.map { $0.description }
+        return text + descriptions.joined(separator: ",\n")
+    }
+}
+
+extension UITouch: CustomStringConvertible {
+    public var description: String {
+        return "TouchID: \(self.touchId), timestamp: \(self.timestamp)"
+    }
+}
+
+#if os(Android)
+import JNI
+
+@_cdecl("Java_org_libsdl_app_SDLActivity_onNativeTouchUIKit")
+public func onNativeTouch(
+    env: UnsafeMutablePointer<JNIEnv>,
+    view: JavaObject,
+    touchDeviceID: JavaInt,
+    pointerFingerID: JavaInt,
+    action: JavaInt,
+    x: JavaFloat,
+    y: JavaFloat,
+    pressure: JavaFloat,
+    timestamp: JavaLong
+) {
+    guard let eventType = SDL_EventType.eventFrom(androidAction: action)
+    else { return }
+
+    var event = SDL_Event(tfinger:
+        SDL_TouchFingerEvent(
+            type: eventType.rawValue,
+            timestamp: UInt32(timestamp), // ensure this is in ms
```

Comment:
these comments are actually TODOs. can you confirm them and if so remove them (would be good to name this input parameter `timestampInMs` if that's the case, for example)